### PR TITLE
Add NonEmptyChain

### PR DESCRIPTION
**Pull request by LukaJCB**
Based on the work in #2371, though we might still need to change a couple of things first, including the `Arbitrary` Generator.

### CODE REVIEW

File: core/src/main/scala/cats/data/NonEmptyChain.scala [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+  def one[A](a: A): NonEmptyChain[A] = create(Chain.one(a))
+
+  implicit def catsNonEmptyChainOps[A](value: NonEmptyChain[A]): NonEmptyChainOps[A] =
+    new NonEmptyChainOps(value)
+}
+
+
+sealed class NonEmptyChainOps[A](val value: NonEmptyChain[A]) {
+
```
Comment (johnynek) [inline core/src/main/scala/cats/data/NonEmptyChain.scala]: Can this extend AnyVal?

File: core/src/main/scala/cats/data/NonEmptyChain.scala [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+  }
+
+  /**
+   * Reduce using the Semigroup of A
+   */
+  final def reduce[AA >: A](implicit S: Semigroup[AA]): AA = {
+    val iter: Iterator[AA] = toChain.iterator
+    var result = iter.next
+    while (iter.hasNext) { result = S.combine(result, iter.next) }
+    result
```
Comment (johnynek) [inline core/src/main/scala/cats/data/NonEmptyChain.scala]: We should use combineAllOption.get so we don’t deoptimize combineAll

**Comment by LukaJCB** [comment]
Urgh, I'm apparently really bad at using git 🤦🏻‍♂️

File: laws/src/main/scala/cats/laws/discipline/Arbitrary.scala [hunk 3]
```diff
@@ -76,6 +76,12 @@ object arbitrary extends ArbitraryInstances0 {
   implicit def catsLawsCogenForNonEmptyList[A](implicit A: Cogen[A]): Cogen[NonEmptyList[A]] =
     Cogen[List[A]].contramap(_.toList)
 
+  implicit def catsLawsArbitraryForNonEmptyChain[A](implicit A: Arbitrary[A]): Arbitrary[NonEmptyChain[A]] =
+    Arbitrary(implicitly[Arbitrary[List[A]]].arbitrary.flatMap(fa => A.arbitrary.map(a => NonEmptyChain(a, fa: _*))))
+
```
Comment (johnynek) [inline laws/src/main/scala/cats/laws/discipline/Arbitrary.scala]: Can we use the Arbitrary for Chain and add one more to the front or back also as a variant?

**Comment by kailuowang** [comment]
@LukaJCB it's not you are bad at using git, it's just the other PR is merged squashed. Since the changes look good, I think if we squash this one as well it will be okay.

File: core/src/main/scala/cats/data/NonEmptyChain.scala [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+      override def exists[A](fa: NonEmptyChain[A])(p: A => Boolean): Boolean =
+        fa.exists(p)
+
+      override def toList[A](fa: NonEmptyChain[A]): List[A] = fa.toChain.toList
+
+      override def toNonEmptyList[A](fa: NonEmptyChain[A]): NonEmptyList[A] =
+        fa.toNonEmptyList
+    }
+
+  implicit def catsDataEqForNonEmptyChain[A: Eq]: Eq[NonEmptyChain[A]] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/NonEmptyChain.scala]: we only have `Eq` here, no Order, PartialOrder, right? Can we add those so we get the priority right there?

File: core/src/main/scala/cats/data/Chain.scala [hunk 5]
```diff
@@ -320,17 +320,17 @@ sealed abstract class Chain[+A] {
   /**
    * Returns the number of elements in this structure
    */
-  final def length: Int = {
+  final def length: Long = {
     val iter = iterator
-    var i: Int = 0
+    var i: Long = 0
```
Comment (kailuowang) [inline core/src/main/scala/cats/data/Chain.scala]: just curious, did we actually tried one instance with more than 2 billion items?

Comment (LukaJCB) [reply core/src/main/scala/cats/data/Chain.scala]: No I don't think so, I just wanted to be consistent with NonEmptyList and NonEmptyVector :)

**Comment by LukaJCB** [comment]
Merging with two signoffs :)

**Pull request merged**